rfd/0136-modern-signature-algorithms.md

```diff
@@ -0,0 +1,876 @@
+will still be allowed to use Ed25519 keys, it is only the CAs that are limited
+to the algorithms supported by the HSM.
+
+### Default suite
+
+For all Teleport clusters where the Auth server's teleport.yaml config file has
+`version: v3` (the current latest version), the default suite will be `legacy`.
+Changing the default will be a breaking change, so we will introduce a config v4
+to change the default.
+
+In config v4 the default suite will be:
+
+* `fips-v1` if the Auth server was started in FIPS mode
+* `hsm-v1` if the Auth server has any HSM or KMS configured (and is not in FIPS mode)
+* `balanced-v1` otherwise
+
+If the suite is explicitly set in the teleport.yaml or
+`cluster_auth_preference`, it will be honoured and no default will be necessary.
+
+### Configuration
+
+The key types and signature algorithms used cluster-wide will be configurable
+via `cluster_auth_preference` and `teleport.yaml`
+
+We want it configurable via `teleport.yaml` so that you can start a new cluster
+and the CA keys will be automatically generated at first start with the correct
+algorithms, so you don't have to immediately edit the `cap` and then rotate all
+of your brand-new CAs.
+
+We want it configurable via `cluster_auth_preference` as well so that it can be
+configurable for Cloud users.
+
+If a value is set in the `cluster_auth_preference`, it will completely override
+the setting from `teleport.yaml`
+
+```yaml
+# teleport.yaml
+version: v3
+teleport:
+  auth_service:
+    enabled: true
+
+    # supported values are balanced-v1, fips-v1, hsm-v1, and legacy
+    signature_algorithm_suite: balanced-v1
+```
+
+```yaml
+kind: cluster_auth_preference
+metadata:
+  name: cluster-auth-preference
+spec:
+  # supported values are balanced-v1, fips-v1, hsm-v1, and legacy
+  signature_algorithm_suite: balanced-v1
+```
+
+### `balanced-v1` suite
+
+The following key types will be used when the configured algorithm suite is
+`balanced-v1`.
+
+* CA key types
+  * user CA
+    * SSH: Ed25519
+    * TLS: ECDSA with NIST P-256
+  * host CA
+    * SSH: Ed25519
+    * TLS: ECDSA with NIST P-256
+  * database CA
+    * TLS: 2048-bit RSA
+      * it's necessary to use RSA for our Snowflake integration, maybe others
+  * OpenSSH CA
+    * SSH: Ed25519
+  * JWT CA
+    * JWT: ECDSA with NIST P-256
+  * OIDC IdP CA
+    * JWT: ECDSA with NIST P-256
+  * SAML IdP CA
+    * TLS: ECDSA with NIST P-256
+* Subject key types
+  * users via `tsh login`
+    * SSH: Ed25519 (SSH cert signed by user CA)
+      * a different algorithm may be used by a hardware key (currently ECDSA with NIST P-256)
+    * TLS: ECDSA with NIST P-256 (X.509 cert signed by user CA)
+  * user web sessions
+    * SSH: Ed25519 (SSH cert signed by user CA)
+    * TLS: ECDSA with NIST P-256 (X.509 cert signed by user CA)
+  * Teleport hosts
+    * SSH: Ed25519 (SSH cert signed by host CA)
+    * TLS: ECDSA with NIST P-256 (X.509 cert signed by host CA)
+  * OpenSSH hosts
```

Comment:
I'm a bit confused by the wording here, I assume `OpenSSH hosts` means OpenSSH host keys? If so, a listing is missing for SSH certs signed by OpenSSH CA when connecting to agentless nodes. You have `SSH: Ed25519 (SSH cert signed by user CA)` above, but that only applies when connecting to Teleport nodes.

Reply:
Thanks, I added a mention for the user certs signed by the OpenSSH CA. `OpenSSH hosts` here refers to both "agentless" and classic openssh nodes, they'll both get Ed25519 SSH keypairs with certs signed by the Host CA.